docs/design/rubygems.md

```diff
@@ -0,0 +1,344 @@
+cat Gemfile.lock
+```
+
+Example of a checksum section in the `Gemfile.lock` file from my custom [repository](https://github.com/slimreaper35/cachi2-rubygems.git):
+
+```Gemfile.lock
+...
+DEPENDENCIES
+  rails (= 6.1.7)
+
+CHECKSUMS
+  actioncable (6.1.7) sha256=ee5345e1ac0a9ec24af8d21d46d6e8d85dd76b28b14ab60929c2da3e7d5bfe64
+  actionmailbox (6.1.7) sha256=c4364381e724b39eee3381e6eb3fdc80f121ac9a53dea3fd9ef687a9040b8a08
+  actionmailer (6.1.7) sha256=5561c298a13e6d43eb71098be366f59be51470358e6e6e49ebaaf43502906fa4
+  actionpack (6.1.7) sha256=3a8580e3721757371328906f953b332d5c95bd56a1e4f344b3fee5d55dc1cf37
+  actiontext (6.1.7) sha256=c5d3af4168619923d0ff661207215face3e03f7a04c083b5d347f190f639798e
+  actionview (6.1.7) sha256=c166e890d2933ffbb6eb2a2eac1b54f03890e33b8b7269503af848db88afc8d4
+  ...
+
+BUNDLED WITH
+   2.5.14
+```
+
+I believe this feature is available since bundler [v2.5.0](https://github.com/rubygems/rubygems/blob/master/bundler/lib/bundler/lockfile_parser.rb#L55)
+from this [PR](https://github.com/rubygems/rubygems/pull/6374) being merged on Oct 21, 2023.
+
+### Environment variables
+
+The order of precedence for bundler configuration options is as follows:
+
+1. Local config (`<project_root>/.bundle/config or $BUNDLE_APP_CONFIG/config`)
+2. Environmental variables (ENV)
+3. Global config (`~/.bundle/config`)
+4. Bundler default config
+
+#### Relevant environment variables
+
+```txt
+BUNDLE_FORCE_RUBY_PLATFORM=true
+BUNDLE_CACHE_ALL=true
+BUNDLE_CACHE_PATH=${output_dir}/deps/rubygems
+BUNDLE_DISABLE_CHECKSUM_VALIDATION=false
+```
+
+**BUNDLE_CACHE_ALL**: Cache all gems, including path and git gems.
+This needs to be explicitly configured on bundler 1 and bundler 2, but will be the default on bundler 3.
+
+**BUNDLE_CACHE_PATH**: The directory that bundler will place cached gems in when running bundle package,
+and that bundler will look in when installing gems. Defaults to `vendor/cache`.
+
+**BUNDLE_FORCE_RUBY_PLATFORM**: Ignore the current machine's platform and install only ruby platform gems.
+As a result, gems with native extensions will be compiled from source.
+
+**BUNDLE_DISABLE_CHECKSUM_VALIDATION**: Allow installing gems even if they do not match the checksum provided by RubyGems.
+
+See bundle config [documentation](https://bundler.io/v2.5/man/bundle-config.1.html).
+
+### Metadata
+
+### git repository URL
+
+- git repository URL is used in other package managers as well
+- no version information available
+- gems in the repository are basically path dependencies in the `Gemfile.lock` ?!
+
+### `{gem}.gemspec` file
+
+- the file is optional / gems in the repository are basically path dependencies in the `Gemfile.lock` ?!
+- complete metadata about the gem
+
+Gemfile must contain a _gemspec_ line + the `{gem}.gemspec` file must be present in the repository.
+Bundle will add the gem as path dependency to the `Gemfile.lock` file.
+
+```Gemfile.lock
+...
+PATH
+  remote: .
+  specs:
+    tmp (0.1.2)
+...
```

Comment:
I'm not sure what you mean by this part. Are you saying that metadata about the "main" gem can be found in Gemfile.lock?

Reply:
Yeah, if the "remote" is `.` it means that there is a gemspec file in the root directory.

I found something similar in the gemlock-parser. Not sure what exactly was meant by [primary gem](https://github.com/containerbuildsystem/gemlock-parser/blob/master/gemlock_parser/gemfile_lock.py#L421)

Reply:
Oh interesting. It sounds like we can use that to solve the metadata parsing https://github.com/containerbuildsystem/cachi2/pull/565#discussion_r1672452499
(and still fall back to git data if there's no gemspec in the root)

Reply:
I found out that it is not mandatory to have the "main" gem entry in the `Gemfile.lock`. Bundler will process it without any issues, and the gemlock-parser will just leave the `primary_gem` attribute as empty.

So I don't think we can rely on this 100%. I'd rather just implement some basic regex parsing for the `.gemspec` file, since that one seems to be required from Bundler's perspective.

Reply:
Here's a [sample file](https://github.com/chatwoot/chatwoot/blob/develop/Gemfile.lock) that does not contain the "main" gem as a PATH dependency.

Reply:
Not mandatory as in which one of these
* the user has to do something special to make sure the main gem appears in Gemfile.lock
* if the user manually removes the main gem from Gemfile.lock, bundler will process the project just fine and put it back
* if the user manually removes the main gem from Gemfile.lock, bundler will not put it back

Reply:
So, for us, this means we can reliably read the `.gemspec` metadata from the `Gemfile`. And if it does not exist, we can fall back to using the Git URL for name, and leave the version empty.

Reply:
I changed the design to reflect the conclusion here.

Reply:
What is exactly wrong with taking it from `Gemfile.lock` and checking if `PATH` == `.` ?
(Even if it is not mandatory)

Reply:
There's nothing wrong with it.

I had the false impression that a `.gemspec` could exist while the `PATH == .` entry in `Gemfile` was absent. That is not what happens, and the only corner case where it could happen is if `gemspec` was not included in the `Gemfile`, which I don't think is Cachi2's problem anyway.